futures/src/supervisor.rs

```diff
@@ -0,0 +1,728 @@
+            Err(Error::UnexpectedExit)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+enum ChildExitResult {
+    Ok,
+    Panic,
+    Cancel,
+}
+
+/// Signal indicating system shutdown. Could be cloned around.
+///
+/// It is effectively a wrap around [`CancellationToken`], but with different naming.
+#[derive(Clone, Debug, Default)]
+pub struct ShutdownSignal(CancellationToken);
+
+impl ShutdownSignal {
+    /// Constructor
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Send the shutdown signal, resolving all [`Self::receive`] futures.
+    pub fn send(&self) {
+        self.0.cancel();
+    }
+
+    /// Receive the shutdown signal. Resolves after [`Self::send`].
+    pub async fn receive(&self) {
+        self.0.cancelled().await
+    }
+
+    /// Sync check whether the shutdown signal was sent
+    pub fn is_sent(&self) -> bool {
+        self.0.is_cancelled()
+    }
+}
+
+/// Spawn [`std::thread`] as a future that finishes when the thread finishes and panics
+/// when the thread panics.
+///
+/// Its intention is to link an OS thread to [`Supervisor`] in the following way:
+///
+/// ```
+/// use std::time::Duration;
+///
+/// use iroha_futures::supervisor::{
+///     spawn_os_thread_as_future, Child, OnShutdown, ShutdownSignal, Supervisor,
+/// };
+///
+/// fn spawn_heavy_work(shutdown_signal: ShutdownSignal) -> Child {
+///     Child::new(
+///         tokio::spawn(spawn_os_thread_as_future(
+///             std::thread::Builder::new().name("heavy_worker".to_owned()),
+///             move || {
+///                 loop {
+///                     if shutdown_signal.is_sent() {
+///                         break;
+///                     }
+///                     // do heavy work...
+///                     std::thread::sleep(Duration::from_millis(100));
+///                 }
+///             },
+///         )),
+///         OnShutdown::Wait(Duration::from(1)),
+///     )
+/// }
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let mut supervisor = Supervisor::new();
+///     supervisor.monitor(spawn_heavy_work(supervisor.shutdown_signal()));
+///
+///     let signal = supervisor.shutdown_signal();
+///     tokio::spawn(async move {
+///         tokio::time::sleep(Duration::from_millis(300)).await;
+///         signal.send();
+///     });
+///
+///     supervisor.start().await.unwrap();
+/// }
+/// ```
+///
+/// **Note:** this function doesn't provide a mechanism to shut down the thread.
+/// You should handle it within the closure on your own, e.g. by passing [`ShutdownSignal`] inside.
+pub async fn spawn_os_thread_as_future<F>(builder: std::thread::Builder, f: F)
+where
+    F: FnOnce(),
+    F: Send + 'static,
+{
+    let (ok_tx, ok_rx) = oneshot::channel();
+    let (err_tx, err_rx) = oneshot::channel();
+
+    // FIXME we cannot just _move_ `err_tx` inside of the thread's panic hook
+    let err_tx = RwLock::new(Some(err_tx));
```

Comment:
Maybe than use `mpsc` and `.blocking_send`?

Reply:
`mpsc` indeed works without a lock (thanks to the ability to have multiple senders).

However, while refactoring, I found that this channel is not needed at all. `ok_rx.await` returns `Err` if the sender is dropped, i.e. if thread panicked. 

Simplified code further: https://github.com/hyperledger/iroha/pull/4716/commits/8a78b8e4faa53b3afadcf25962ee668139d23eaf